lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -4094,6 +4017,281 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
+  auto *original = getOriginal();
+  // Parameters of the adjoint are:
+  // - a seed,
+  // - a primal value struct,
+  // - original results, and
+  // - the original parameters.
+  // Results of the adjoint have the same type as the original parameters.
+  SmallVector<SILParameterInfo, 8> adjParams;
+  SmallVector<SILResultInfo, 8> adjResults;
+  auto origTy = original->getLoweredFunctionType();
+  auto origParams = origTy->getParameters();
+
+  // Add adjoint parameter for the seed.
+  adjParams.push_back(getFormalParameterInfo(
+      origTy->getResults()[getIndices().source].getType(), module));
+
+  // If there's a generated primal, accept a primal value struct in the adjoint
+  // parameter list.
+  if (auto *pi = getPrimalInfo()) {
+    auto pvType = pi->getPrimalValueStruct()
+                      ->getDeclaredInterfaceType()
+                      ->getCanonicalType();
+    adjParams.push_back(getFormalParameterInfo(pvType, module));
+  }
+
+  // Add adjoint parameters for the original results.
+  for (auto &origRes : origTy->getResults())
+    adjParams.push_back(getFormalParameterInfo(origRes.getType(), module));
+
+  // Add adjoint parameters for the original parameters.
+  for (auto &param : origParams)
+    adjParams.push_back(param);
+
+  // Add adjoint result for the wrt self parameter, if it was requested.
+  auto selfParamIndex = origParams.size() - 1;
+  if (origTy->hasSelfParam() &&
+      getIndices().isWrtParameter(selfParamIndex))
+    adjResults.push_back(getFormalResultInfo(
+        origParams[selfParamIndex].getType(), module));
+
+  // Add adjoint results for the requested non-self wrt parameters.
+  for (auto i : getIndices().parameters.set_bits()) {
+    if (origTy->hasSelfParam() && i == selfParamIndex)
+      continue;
+    adjResults.push_back(getFormalResultInfo(origParams[i].getType(), module));
+  }
+
+  auto adjName = "AD__" + original->getName().str() + "__adjoint_" +
+                 mangleADIndices(getIndices());
+  auto adjType = SILFunctionType::get(
+      origTy->getGenericSignature(), origTy->getExtInfo(),
+      origTy->getCoroutineKind(), origTy->getCalleeConvention(), adjParams, {},
+      adjResults, None, original->getASTContext());
+  SILOptFunctionBuilder fb(context.getTransform());
+  auto linkage = original->getLinkage();
+  if (linkage == SILLinkage::Public)
+    linkage = SILLinkage::PublicNonABI;
+  adjoint = fb.createFunction(linkage, adjName, adjType,
+      original->getGenericEnvironment(), original->getLocation(),
+      original->isBare(), original->isTransparent(), original->isSerialized());
+  adjoint->setUnqualifiedOwnership();
+  adjoint->setDebugScope(new (module)
+                             SILDebugScope(original->getLocation(), adjoint));
+
+  attr->setAdjointName(adjoint->getName(), /*primitive*/ false);
+}
+
+void DifferentiationTask::createVJP() {
+  // FIXME: Support generics.
+  if (getOriginal()->getLoweredFunctionType()->getGenericSignature()) {
+    return;
```

Comment:
We don't have the ability to do PrimalGen or AdjointGen for generic functions yet. How is the VJP synthesized?

Reply:
Discussed in person.

For anyone reading: The next PR (https://github.com/apple/swift/pull/21062) can create the empty shells of generic primals and adjoints, but it still bails out when PrimalGen tries to start filling in the primal.

The next PR can also create a full generic VJP with body that calls primal and adjoint. So if the primal and adjoint for a generic function are manually specified, then a fully-working VJP gets created. This is just a convenience so that I can test that VJPs work on code that has custom primal/adjoint and no custom VJP. Soon this convenience will become irrelevant because we will forbid custom primal/adjoint and replace them with custom VJP.